org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/pom/ReImportMavenProjectsHandler.java

```diff
@@ -136,36 +95,42 @@
   }
 
   /**
-   * Submits jobs to update given maven projects.
+   * Updates maven projects.
    *
    * @param projects list of projects to be updated
+   * @return list of jobs to update projects. It is needed for tests
    */
-  private static void submitUpdateJobs(List<IProject> projects) {
+  public static List<Job> updateProjects(List<IProject> projects) {
+    List<Job> updatedJobs = new ArrayList<>();
     for (IProject project : projects) {
-      projectsManager.updateProject(project, true);
+      Job job = JavaLanguageServerPlugin.getProjectsManager().updateProject(project, true);
+      job.addJobChangeListener(new JobChangedListener());
+      updatedJobs.add(job);
     }
+    return updatedJobs;
   }
 
-  // TODO temporary solution.
-  // Should be reworked when https://github.com/eclipse/eclipse.jdt.ls/issues/506 will be fixed.
   private static class JobChangedListener extends JobChangeAdapter {
     private static final String UPDATE_PROJECT_JOB_NAME_PREFIX = "Update project ";
 
     /** Listens to jobs changes events to get information about updated projects. */
     @Override
     public void done(IJobChangeEvent event) {
       String jobName = event.getJob().getName();
-      if (jobName.startsWith(UPDATE_PROJECT_JOB_NAME_PREFIX)) {
-        String projectName = jobName.substring(UPDATE_PROJECT_JOB_NAME_PREFIX.length());
-        if (projectsToBeUpdated.containsKey(projectName)) {
-          if (event.getResult().isOK()) {
-            updatedProjectsPaths.put(projectsToBeUpdated.remove(projectName), "");
-          } else {
-            projectsToBeUpdated.remove(projectName);
-          }
-
-          jobsToBeFinished.countDown();
-        }
+      if (!jobName.startsWith(UPDATE_PROJECT_JOB_NAME_PREFIX)) {
```

Comment:
There is no reason to match jobs by name. We _know_ every single job we need to wait for, because we create them on line 106 above. Please get rid of the jobs name matching logic and add specific job change listeners above.

Reply:
fixed